lib/widgets/login.dart

```diff
@@ -27,27 +107,57 @@ class AddAccountPage extends StatefulWidget {
 }
 
 class _AddAccountPageState extends State<AddAccountPage> {
-  final TextEditingController _controller = TextEditingController();
+  bool _inProgress = false;
+
+  final ServerUrlTextEditingController _controller = ServerUrlTextEditingController();
+  late ServerUrlParseResult _parseResult;
+
+  _serverUrlChanged() {
+    setState(() {
+      _parseResult = _controller.tryParse();
+    });
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _parseResult = _controller.tryParse();
+    _controller.addListener(_serverUrlChanged);
+  }
 
   @override
   void dispose() {
     _controller.dispose();
     super.dispose();
   }
 
-  Future<void> _onSubmitted(BuildContext context, String value) async {
-    final Uri? url = Uri.tryParse(value);
-    switch (url) {
-      case Uri(scheme: 'https' || 'http'):
-        // TODO(#35): validate realm URL further?
-        break;
-      default:
-        // TODO(#35): give feedback to user on bad realm URL
-        return;
+  Future<void> _onSubmitted(BuildContext context) async {
+    final ServerUrlParseResult(:url, :error) = _parseResult;
+    if (error != null) {
+      showErrorDialog(context: context,
+        title: 'Invalid input', message: error.message());
+      return;
     }
+    assert(url != null);
 
-    // TODO(#35): show feedback that we're working, while fetching server settings
-    final serverSettings = await getServerSettings(realmUrl: url);
+    final GetServerSettingsResult serverSettings;
+    setState(() {
+      _inProgress = true;
+    });
+    try {
+      serverSettings = await getServerSettings(realmUrl: url!);
+    } catch (e) {
+      if (!context.mounted) return;
+      // TODO(#35) give more helpful feedback; see `fetchServerSettings`
+      //   in zulip-mobile's src/message/fetchActions.js. Needs #37.
+      showErrorDialog(context: context,
+        title: 'Could not connect', message: 'Failed to connect to server: $url');
```

Comment:
```suggestion
        title: 'Could not connect', message: 'Failed to connect to server:\n$url');
```

That helps avoid wrapping the URL.